File: 1. JPA basic/basic/src/main/java/study/jpa/basic/jpashop/domain/Delivery.java
package study.jpa.basic.jpashop.domain;

import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.OneToOne;
import javax.persistence.Table;

/**
 * @author Joonhyuck Hyoung
 */
@Entity(name = "jpashop_delivery")
@Table(name = "jpashop_delivery")
public class Delivery extends BaseEntity {
    @Id @GeneratedValue
    private Long no;

    @Embedded
    private Address address;

    private String city;
    private String street;
    private String zipCode;
    private DeliveryStatus status;

    // 양방향
    @OneToOne(mappedBy = "delivery", fetch = FetchType.LAZY)
    private Order order;

    public Address getAddress() {
        return address;
    }

    public void setAddress(Address address) {
        this.address = address;
    }
}

File: 2. JPA deep/jpashop/src/main/kotlin/jpabook/jpashop/domain/Delivery.kt
package jpabook.jpashop.domain

import net.minidev.json.annotate.JsonIgnore
import javax.persistence.*

@Entity
data class Delivery(
    @Id @GeneratedValue
    @Column(name = "delivery_id")
    val id: Long = 0L,

    @JsonIgnore
    @OneToOne(mappedBy = "delivery", fetch = FetchType.LAZY)
    var order: Order? = null,

    @Embedded
    val address: Address? = null,

    @Enumerated(EnumType.STRING)
    val status: DeliveryStatus = DeliveryStatus.READY,
) {
    companion object {
        fun create(address: Address?, status: DeliveryStatus): Delivery {
            return Delivery(
                address = address,
                status = status
            )
        }
    }
}
